specs/emerald_app.qnt

```diff
@@ -173,528 +116,277 @@ module emerald_app {
   // HELPER FUNCTIONS
   // =============================================================================
 
-  // Check if a proposal exists in a given status
-  pure def has_proposal(proposals: Set[Proposal], h: Height, r: Round, status: ProposalStatus): bool = {
-    proposals.exists(p => p.height == h and p.round == r and p.status == status)
-  }
-
-  // Get proposal by height, round, and value_id
-  pure def get_proposal(proposals: Set[Proposal], h: Height, r: Round, vid: ValueId): Option[Proposal] = {
-    proposals.filter(p => p.height == h and p.round == r and p.value_id == vid).fold(None, (_, p) => Some(p))
-  }
+  pure def proposer_for(height: Height, round: Round): Node =
+    NODES[(height - 1 + round) % NODES.length()]
```

Comment:
We don't need to change this now, but I think I prefer the usual way of modeling proposers as a map of round to node (or, in this case, I guess a mix of that with some circular thing for heights. We usually don't model different heights). The map allows us to more easily manipulate things to make byzantine nodes proposers of certain rounds. Maybe the same level of manipulation can be easily achieved here by making specific nodes behave as byzantines, or setting `NODES` to something like `["byz1", "node1", ...]` to make a byzantine propose on the first round, first height.

As I typed this, I realized this solution might actually be good, we should at least experiment. So let's keep it and see how things go with this in mind.